### PR TITLE
Fixes #528,464: Improvements and bug fixes in processing the INSTANCENAME parameter

### LINKED ISSUE #528: Expand help on errors for invalid wbem_url
The exception for wbem uri leaves a lot to be desired, in particular, more information on what should be in it and what the error might be.

We need more information in the exception.  See the code in _cmd_instances.py function
get_instancename().

### PR DESCRIPTION
**Pull request by andy-maier**
Ready for review. No longer based on PR #552
For details, see commit message.

Good manual tests are to specify all kinds of invalid INSTANCENAME, --key and --namespace options and to verify whether the error messages make sense.

### CODE REVIEW

**Comment by andy-maier** [comment]
**Discussion point**:  `get_instancename()` performs a GetClass when the keybindings are specified with the `--key` option. It would be beneficial to avoid the GetClass, because GetClass is not implemented by all WBEM servers.

That is not easy however, because the `name=value` syntax used in the `--key` option does not allow distinguishing the string/numeric/boolean types sufficiently well. For example, one would need to find a way to resolve the ambiguity between numeric/boolean values and strings with that same string value.

One option might be to have optional single or double quotes around string values, that normally would not be needed, but when the string value is a numeric or boolean string, then they would be used to turn the value into a string.

For example:
* `name=42` -> numeric, `name="42"` -> string
* `name=true` -> boolean, `name="true"` -> string

This obviously complicates the syntax and needs to be described, including how the quotes are treated by shells (to be verified, but probably not).

Discussion point is should we introduce that?

**Comment by KSchopmeyer** [comment]
We agreed to Andy's proposal.

**Comment by andy-maier** [comment]
The PR is ready for review. Each of its 3 commits is a logical change on its own.

**Comment by andy-maier** [comment]
Rebased after merging PR #552.

File: pywbemtools/pywbemcli/_cmd_instance.py [hunk 1]
```diff
@@ -674,6 +639,122 @@ class are displayed and the user is prompted for an index number to select
...
+   approach and eliminates it for the most common cases, but the processing of
+   special characters by the shell still needs to be considered.
+
+   The --key option can be specified multiple times, and the option argument
+   of each use has a NAME=VALUE format, where NAME is the name of the key
+   property and VALUE is its value. The string/numeric/boolean type needed for
+   creating a keybinding is determined automatically from VALUE. Valid integer
+   numbers are interpreted as a numeric type, the strings "true" and "false" in
+   any lexical case are interpreted as a boolean type, and anything else is
+   interpreted as a string type. If VALUE starts and ends with double quotes,
```
[the remark this answers is not in the document]
Comment (andy-maier) [reply pywbemtools/pywbemcli/_cmd_instance.py:710]: DONE

File: docs/pywbemcli/cmdshelp.rst [hunk 2]
```diff
@@ -995,6 +986,9 @@ Help text for ``pywbemcli instance associators`` (see :ref:`instance associators
 
       --fql, --filter-query-language QUERY-LANGUAGE
                                       The filter query language to be used with --filter-query. Default: DMTF:FQL.
+      --hi, --help-instancename       Show help message for specifying INSTANCENAME including use of the --key and
+                                      --namespace options and exit.
```
[the remark this answers is not in the document]
Comment (andy-maier) [reply docs/pywbemcli/cmdshelp.rst:990]: I agree, we should just leave it off. Everybody expects that nothing else happens besides showing the help text.

File: pywbemtools/pywbemcli/_cmd_instance.py [hunk 3]
```diff
@@ -674,6 +639,122 @@ class are displayed and the user is prompted for an index number to select
...
+
+2. By specifying the class path of the creation class of the instance as an
+   untyped WBEM URI in the INSTANCENAME argument and by using the --key option
+   to specify the keybindings of the instance.
+
+   This approach reduces the use of double quotes compared to the first
+   approach and eliminates it for the most common cases, but the processing of
+   special characters by the shell still needs to be considered.
+
+   The --key option can be specified multiple times, and the option argument
```
[the remark this answers is not in the document]
Comment (andy-maier) [reply pywbemtools/pywbemcli/_cmd_instance.py:704]: Added ", once for each key of the instance name". DONE

File: pywbemtools/pywbemcli/_cmd_instance.py [hunk 4]
```diff
@@ -674,6 +639,122 @@ class are displayed and the user is prompted for an index number to select
...
+     'TST_Foo.StringKey="1.75\\""'      # shell processes outer squotes
+     'TST_Foo.StringKey="a b"'         # shell processes outer squotes
+     'CIM_SubProfile.Main="CIM_RegisteredProfile.InstanceID=\\"acme.1\\"",
+       Sub="CIM_RegisteredProfile.InstanceID=\\"acme.2\\""'
+
+2. By specifying the class path of the creation class of the instance as an
+   untyped WBEM URI in the INSTANCENAME argument and by using the --key option
+   to specify the keybindings of the instance.
+
+   This approach reduces the use of double quotes compared to the first
```
[the remark this answers is not in the document]
Comment (andy-maier) [reply pywbemtools/pywbemcli/_cmd_instance.py:700]: We don't get information on the property type from the server anymore, and we don't need that anymore. The approach with the double quotes to force string interpretation is needed only in rare circumstances, i.e. when the string value is a decimal number, or "true", or "false", and it is very natural and last but not least it was supported already before this PR.

**Comment by andy-maier** [comment]
One discussion needed left, and one comment will be resolved in a separate PR.

File: pywbemtools/pywbemcli/_cmd_instance.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (andy-maier) [reply pywbemtools/pywbemcli/_cmd_instance.py:700]: accepted.

**Comment by andy-maier** [comment]
Karl wants to review the code.

**Pull request merged**